README
NAME
    PYX::Sort - Processing PYX data or file and sort element attributes.

SYNOPSIS
     use PYX::Sort;

     my $obj = PYX::Sort->new(%parameters);
     $obj->parse($pyx, $out);
     $obj->parse_file($input_file, $out);
     $obj->parse_handler($input_file_handler, $out);

METHODS
  "new"
     my $obj = PYX::Sort->new(%parameters);

    Constructor.

    *       "output_handler"

            Output handler.

            Default value is \*STDOUT.

    Returns instance of object.

  "parse"
     $obj->parse($pyx, $out);

    Parse PYX text or array of PYX text and print sorted list of element
    attributes in PYX format. If $out not present, use 'output_handler'.

    Returns undef.

  "parse_file"
     $obj->parse_file($input_file, $out);

    Parse file with PYX data and print sorted list of element attributes in
    PYX format. If $out not present, use 'output_handler'.

    Returns undef.

  "parse_handler"
     $obj->parse_handler($input_file_handler, $out);

    Parse PYX handler print sorted list of element attributes in PYX format.
    If $out not present, use 'output_handler'.

    Returns undef.

ERRORS
     new():
             From Class::Utils::set_params():
                     Unknown parameter '%s'.

EXAMPLE
     use strict;
     use warnings;

     use PYX::Sort;

     # Example data.
     my $pyx = <<'END';
     (tag
     Aattr3 value
     Aattr2 value
     Aattr1 value
     -text
     )tag
     END

     # PYX::Sort object.
     my $obj = PYX::Sort->new;

     # Parse.
     $obj->parse($pyx);

     # Output:
     # (tag
     # Aattr1="value"
     # Aattr2="value"
     # Aattr3="value"
     # -text
     # )tag

DEPENDENCIES
    Class::Utils, PYX::Parser.

SEE ALSO
    Task::PYX
        Install the PYX modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/PYX-Sort>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2011-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.06



Changes
0.06

0.05 2023-10-19T10:51:00+02:00
 - Fix typo in doc.
 - Rewrite doc to use =head2. Improve doc.

0.04 2023-09-11T18:14:08+02:00
 - Fix bugtracker.
 - Fix sense of example in doc.
 - Fix use lib '.' in Makefile.PL.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Rename example to better name.
 - Update Module::Install to 1.21 version.
 - Update author github username.
 - Update author username.
 - Update copyright years.

0.03 2015-10-21T15:02:50+02:00
 - Add tests for parse() and parse_handler() methods.
 - Improve SEE ALSO section in doc.
 - Move author tests to xt/ directory.
 - Update Module::Install to 1.16 version.

0.02 2015-03-22T01:14:29+01:00
 - Fix dependencies for PYX::Parser.

0.01 2015-03-20T19:05:32+01:00
 - First version.


Sort.pm
package PYX::Sort;

use strict;
use warnings;

use Class::Utils qw(set_params);
use PYX::Parser;

our $VERSION = 0.06;

# Constructor.
sub new {
	my ($class, @params) = @_;
	my $self = bless {}, $class;

	# Output handler.
	$self->{'output_handler'} = \*STDOUT;

	# Process params.
	set_params($self, @params);

	# PYX::Parser object.
	$self->{'pyx_parser'} = PYX::Parser->new(
		'callbacks' => {
			'attribute' => \&_attribute,
			'start_element' => \&_tag,
			'end_element' => \&_tag,
			'comment' => \&_tag,
			'instruction' => \&_tag,
			'data' => \&_tag,
		},
		'non_parser_options' => {
			'tag' => {},
		},
		'output_handler' => $self->{'output_handler'},
	);

	# Object.
	return $self;
}

# Parse pyx text or array of pyx text.
sub parse {
	my ($self, $pyx, $out) = @_;
	$self->{'pyx_parser'}->parse($pyx, $out);
	return;
}

# Parse file with pyx text.
sub parse_file {
	my ($self, $file, $out) = @_;
	$self->{'pyx_parser'}->parse_file($file, $out);
	return;
}

# Parse from handler.
sub parse_handler {
	my ($self, $input_file_handler, $out) = @_;
	$self->{'pyx_parser'}->parse_handler($input_file_handler, $out);
	return;
}

# Process attribute.
sub _attribute {
	my ($pyx_parser_obj, $att, $attval) = @_;
	$pyx_parser_obj->{'non_parser_options'}->{'tag'}->{$att} = $attval;
	return;
}

# Process tag.
sub _tag {
	my $pyx_parser_obj = shift;
	my $out = $pyx_parser_obj->{'output_handler'};
	_flush($pyx_parser_obj);
	print {$out} $pyx_parser_obj->line, "\n";
	return;
}

# Flush attributes.
sub _flush {
	my $pyx_parser_obj = shift;
	my $out = $pyx_parser_obj->{'output_handler'};
	foreach my $key (sort keys %{$pyx_parser_obj->{'non_parser_options'}
		->{'tag'}}) {

		print {$out} 'A'.$key.'="'.$pyx_parser_obj
			->{'non_parser_options'}->{'tag'}->{$key}.'"'."\n";
	}
	$pyx_parser_obj->{'non_parser_options'}->{'tag'} = {};
	return;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

PYX::Sort - Processing PYX data or file and sort element attributes.

=head1 SYNOPSIS

 use PYX::Sort;

 my $obj = PYX::Sort->new(%parameters);
 $obj->parse($pyx, $out);
 $obj->parse_file($input_file, $out);
 $obj->parse_handler($input_file_handler, $out);

=head1 METHODS

=head2 C<new>

 my $obj = PYX::Sort->new(%parameters);

Constructor.

=over 8

=item * C<output_handler>

Output handler.

Default value is \*STDOUT.

=back

Returns instance of object.

=head2 C<parse>

 $obj->parse($pyx, $out);

Parse PYX text or array of PYX text and print sorted list of element attributes in PYX format.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<parse_file>

 $obj->parse_file($input_file, $out);

Parse file with PYX data and print sorted list of element attributes in PYX format.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<parse_handler>

 $obj->parse_handler($input_file_handler, $out);

Parse PYX handler print sorted list of element attributes in PYX format.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head1 ERRORS

 new():
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

=head1 EXAMPLE

=for comment filename=sort_element_attributes.pl

 use strict;
 use warnings;

 use PYX::Sort;

 # Example data.
 my $pyx = <<'END';
 (tag
 Aattr3 value
 Aattr2 value
 Aattr1 value
 -text
 )tag
 END

 # PYX::Sort object.
 my $obj = PYX::Sort->new;

 # Parse.
 $obj->parse($pyx);

 # Output:
 # (tag
 # Aattr1="value"
 # Aattr2="value"
 # Aattr3="value"
 # -text
 # )tag

=head1 DEPENDENCIES

L<Class::Utils>,
L<PYX::Parser>.

=head1 SEE ALSO

=over

=item L<Task::PYX>

Install the PYX modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/PYX-Sort>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2011-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.06

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Processing PYX data or file and sort element attributes.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.07;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'PYX-Sort';
readme_from 'Sort.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.05;
requires 'PYX::Parser' => 0.03;
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/PYX-Sort/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/PYX-Sort';
resources 'repository' => 'git://github.com/michal-josef-spacek/PYX-Sort';
test_requires 'English' => 0;
test_requires 'File::Object' => 0.07;
test_requires 'Perl6::Slurp' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
test_requires 'Test::Output' => '1.02';
tests_recursive;
version '0.06';

# Run.
WriteAll();
